src/libraries/System.Formats.Nrbf/src/System/Formats/Nrbf/MemberTypeInfo.cs

```diff
@@ -53,10 +53,14 @@ internal static MemberTypeInfo Decode(BinaryReader reader, int count, PayloadOpt
                 case BinaryType.Class:
                     info[i] = (type, ClassTypeInfo.Decode(reader, options, recordMap));
                     break;
-                default:
-                    // Other types have no additional data.
-                    Debug.Assert(type is BinaryType.String or BinaryType.ObjectArray or BinaryType.StringArray or BinaryType.Object);
+                case BinaryType.String:
+                case BinaryType.StringArray:
+                case BinaryType.Object:
+                case BinaryType.ObjectArray:
+                    // These types have no additional data.
                     break;
+                default:
+                    throw new InvalidOperationException();
```

Comment:
FWIW, InvalidOperationException is for when a method is invoked on an object and the object has state that doesn't support it.  For example, calling Stream.Read when Stream.CanRead says false.  The object is not in a state where you can call the method.

Many of yours are probably more likely InvalidDataException, where the default message is "Found invalid data while decoding."

Reply:
Immo corrected me.  Stream.Read when Stream.CanRead is false is NotSupportedException (you should have checked the property on the stream).  InvalidOperationException is for controllable state that was controlled wrong, like an object where you have to set 3 properties before calling "DoStuff", and you called DoStuff without setting all 3.

But this is still most likely InvalidDataException, because the operation made sense to do, just the data that it read back was not internally consistent.

Reply:
What I wanted to achieve is:
- ensure all methods that parse enums throw SerializationException when they encounter invalid value (out of range, not in the spec)
- assume that the above is not true and always handle all enum values in switch statements in explicit way, throw in the default case (so if there is a bug in parsing logic, we don't get undefined behavior but just an exception) 

My initial idea was to use `UnreachableException`, but it's not part of NS2.0. I decided to use `InvalidOperationException` and I am open to changing it.

It's being tested by the Fuzzer which does not catch `InvalidOperationException` and treats it as a bug:

https://github.com/dotnet/runtime/blob/4930e1b44027484e13a29c7f3230e1f0bb397e7c/src/libraries/Fuzzing/DotnetFuzzing/Fuzzers/NrbfDecoderFuzzer.cs#L100-L106

(as shown in the fuzzer comments in this PR that has discovered places where it was missing)